pkg/auth/providers/local/pbkdf2/pbkdf2.go

```diff
@@ -0,0 +1,222 @@
+package pbkdf2
+
+import (
+	"bytes"
+	"crypto/pbkdf2"
+	"crypto/rand"
+	"crypto/sha3"
+	"encoding/json"
+	"fmt"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/wrangler/v3/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/crypto/bcrypt"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	LocalUserPasswordsNamespace = "cattle-local-user-passwords"
+	iterations                  = 210000
+	keyLength                   = 32
+	passwordHashAnnotation      = "cattle.io/password-hash"
+	pbkdf2sha3512Hash           = "pbkdf2sha3512"
+	bcryptHash                  = "bcrypt"
+)
+
+// Pbkdf2 handles password storage and hashing using PBKDF2.
+type Pbkdf2 struct {
+	secretLister  v1.SecretCache
+	secretClient  v1.SecretClient
+	hashKey       func(password string, salt []byte, iter, keyLength int) ([]byte, error)
+	saltGenerator func() ([]byte, error)
+}
+
+func New(secretLister v1.SecretCache, secretClient v1.SecretClient) *Pbkdf2 {
+	return &Pbkdf2{
+		secretLister:  secretLister,
+		secretClient:  secretClient,
+		hashKey:       sha3512Key,
+		saltGenerator: generateSalt,
+	}
+}
+
+// CreatePassword hashes the provided password using PBKDF2 and stores it in a secret associated with the specified user.
+func (p *Pbkdf2) CreatePassword(user *v3.User, password string) error {
+	salt, err := p.saltGenerator()
+	if err != nil {
+		return fmt.Errorf("failed to generate salt: %w", err)
+	}
+	hashedPassword, err := p.hashKey(password, salt, iterations, keyLength)
+	if err != nil {
+		return fmt.Errorf("failed to hash password: %w", err)
+	}
+
+	_, err = p.secretClient.Create(&corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      user.Name,
+			Namespace: LocalUserPasswordsNamespace,
+			Annotations: map[string]string{
+				passwordHashAnnotation: pbkdf2sha3512Hash,
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name:       user.Name,
+					UID:        user.UID,
+					APIVersion: "management.cattle.io/v3",
+					Kind:       "User",
+				},
+			},
+		},
+		Data: map[string][]byte{
+			"password": hashedPassword,
+			"salt":     salt,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create secret: %w", err)
+	}
+
+	return nil
+}
+
+// UpdatePassword hashes the provided password using PBKDF2 and updates the secret associated with the specified user
+func (p *Pbkdf2) UpdatePassword(userId string, newPassword string) error {
+	secret, err := p.secretLister.Get(LocalUserPasswordsNamespace, userId)
+	if err != nil {
+		return fmt.Errorf("failed to get password secret: %w", err)
+	}
+
+	hashedNewPassword, err := p.hashKey(newPassword, secret.Data["salt"], iterations, keyLength)
```

Comment:
Query: Any specific reason for the new password using the salt of the old password ?
Would using a new salt be better ?